### ISSUE TITLE
Question about generating training dataset

### ISSUE THREAD

**Issue by fenneccat** [lpworld/PURS#9]
```
train_user = train_data.loc[train_data['user_id']==user]
    train_user = train_user.sort_values(['hour'])
    length = len(train_user)
    train_user.index = range(length)
    if length > 10:
        for i in range(length-10):
            train_set.append((train_user.loc[i+9,'user_id'], list(train_user.loc[i:i+9,'video_id']), train_user.loc[i+9,'video_id'], float(train_user.loc[i+9,'click'])))
```

According to the above code, train_set contains, user_id, history, recommended item, and click-or-non click label.

I have two questions here

1. What I'm curious about is it seems history's last item is the same as the recommended item.
I thought the recommended item to predict CTR would be totally new, but it wasn't. Can you tell me a reason why history's last element and the recommended item are the same?

2. As I understood, history is the watching-list of users. However, when making history list, `list(train_user.loc[i:i+9,'video_id'])` it seems it doesn't care whether the user watched this video or not. Since test.txt contains negative samples which have click label as zero, I thought history should contain only videos which are clicked and omit instances which don't clicked. Can you explain why training data is designed like this?

**Comment by lpworld** [comment]
(1) No, it's not the same. Considering the following example: history_sequence = [1,2,3,4,5,6,7,8,9,10]; history_sequence[0:9] = [1,2,3,4,5,6,7,8,9]; history_sequence[9] = [10]

(2) This is a good question. As our recommendation setting is the video streaming feed (like TikTok), users would get exposed to the recommended videos anyway (either clicked or non-clicked), so we do not distinguish the positive samples and the negative samples in our setting. Of course, you can single out the positive samples to compute unexpectedness for other recommendation tasks.

**Comment by fenneccat** [comment]
(1) according to dataframe.loc document (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html), it seems df.loc[a:b] both include a and b. When I print the data, it was true. That's why I asked you.

(2) I see. So you set history as impressions on video streaming feed. Thank you!

**Comment by uvj1010** [comment]
Regarding Chapter 3.1 (Modeling of Unexpectedness) in the paper, historic behavior sequence (length = n) will be clustered into N user interest clusters, and for a new item, unexp is the weighted average distance between each cluster and
embedding of the new item.

But in the code (train.py - mean_shift) :

1. Each historical item is initialized as a center and iteratively shifted. The number of centers output is equal to the length of the historical sequence (n = N).

2. When calculating the distance between new item and clusters, all centers are averaged into one center, which is not consistent with the "weighted average" and "multi user interest clusters" represented by formula 3.

I think this implementation method is better in terms of calculation speed, and the dimension is fixed if the center is not merged, but does this calculation result meet expectations?
